### PR TITLE
TECH-509: Add endpoint to get bulding blocks list (BE)

### PR DESCRIPTION
**Pull request by mgrochalskisoldevelo**
## PR description
Created an endpoint to fetch building blocks list with count of their supported softwares. 

Response preview: 
image

## Ticket description
Add endpoint to get the list of existing building blocks. It should include:
- Building block’s ID
- Number of supported softwares

Acceptance Criteria
- Send data with the list of building blocks available and the number of supported softwares

TICKET: https://govstack-global.atlassian.net/browse/TECH-509

### CODE REVIEW

File: server/src/db/repositories/reportRepository.js [hunk 1]
```diff
@@ -30,6 +31,10 @@ const repository = () => {
     ReportModel.aggregate(sortedAggregation).exec(callback);
   };
 
+  const aggregateBuildingBlocks = (callback) => {
+    ReportModel.aggregate(getBuildingBlockListPipeline()).exec(callback);
```
Comment (dborowiecki) [inline server/src/db/repositories/reportRepository.js:35]: Just a note but we should not rely on pipelines in every case. For compatibility aggregations it's unavoidable but Gettling list of unique bbs and relevant software seems more simplistic and perhaps we should put more focus on utilization of [mongoose queries](https://mongoosejs.com/docs/api/query.html) or [aggregations ](https://mongoosejs.com/docs/api/aggregate.html). Which are much more readable than json schema pipelines.

**Review by dborowiecki — CHANGES_REQUESTED** [review]
Please reefer to my comments

File: server/src/db/pipelines/buildingBlocks.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
+const getBuildingBlockListPipeline = () => [
+  {
+    $match: {
```
[the remark this answers is not in the document]
Comment (mgrochalskisoldevelo) [reply server/src/db/pipelines/buildingBlocks.js]: I removed the matching criteria, do you think it should be added and separated, or we can let it be like that.

File: server/src/controllers/reportController.js [hunk 3]
```diff
@@ -2,6 +2,7 @@ const ReportUploadRequestHandler = require('../useCases/report/reportSave/handle
 const ReportProductGetRequestHandler = require('../useCases/report/productCompatibility/handleGetProductRequest');
 const ProductCountRequestHandler = require('../useCases/report/productCount/handleGetProductCountRequest');
 const ProductDetailsRequestHandler = require('../useCases/report/productTestCases/handleGetProductDetailsRequest');
+const BuildingBlocksRequestHandler = require('../useCases/report/buildingBlocks/handleGetBuildingBlocksRequest');
```
[the remark this answers is not in the document]
Comment (mgrochalskisoldevelo) [reply server/src/controllers/reportController.js]: changed to ```ReportGetBuildingBlocksRequestHandler```

File: server/src/db/repositories/reportRepository.js [hunk 4]
```diff
@@ -30,6 +31,10 @@ const repository = () => {
     ReportModel.aggregate(sortedAggregation).exec(callback);
   };
 
+  const aggregateBuildingBlocks = (callback) => {
```
[the remark this answers is not in the document]
Comment (mgrochalskisoldevelo) [reply server/src/db/repositories/reportRepository.js]: I changed it to ```aggregateByBuildingBlock``` but haven't passed any additional parameter. Is there something that should be passed?

File: server/src/db/repositories/reportRepository.js (hunk 1 above)
Comment (mgrochalskisoldevelo) [reply server/src/db/repositories/reportRepository.js]: changed it to 
```
    const aggregation = ReportModel.aggregate([]);
    aggregation.group({
      _id: '$buildingBlock',
      softwares: { $addToSet: '$testApp' },
    });
    aggregation.project({
      _id: 1,
      supportedSoftwares: { $size: '$softwares' },
    });
    aggregation.sort({ _id: 1 }).exec(callback);
```
is that what you meant?

**Review by dborowiecki — APPROVED** [review]

**Pull request merged**